zingolib/src/wallet/transaction_records_by_id.rs

```diff
@@ -21,8 +31,94 @@ impl std::ops::DerefMut for TransactionRecordsById {
     }
 }
 impl TransactionRecordsById {
+    // initializers
+
     // Associated function to create a TransactionRecordMap from a HashMap
+    pub fn default() -> Self {
+        TransactionRecordsById(HashMap::new())
+    }
     pub fn from_map(map: HashMap<TxId, TransactionRecord>) -> Self {
         TransactionRecordsById(map)
     }
+
+    // modify methods
+
+    pub fn invalidate_all_transactions_after_or_at_height(&mut self, reorg_height: BlockHeight) {
+        // First, collect txids that need to be removed
+        let txids_to_remove = self
+            .values()
+            .filter_map(|transaction_metadata| {
+                if transaction_metadata
+                    .status
+                    .is_confirmed_after_or_at(&reorg_height)
+                    || transaction_metadata
+                        .status
+                        .is_broadcast_after_or_at(&reorg_height)
+                // tODo: why dont we only remove confirmed transactions. unconfirmed transactions may still be valid in the mempool and may later confirm or expire.
+                {
+                    Some(transaction_metadata.txid)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        self.invalidate_txids(txids_to_remove);
+    }
+
+    /// this function invalidiates a vec of txids
+    pub(crate) fn invalidate_txids(&mut self, txids_to_remove: Vec<TxId>) {
+        for txid in &txids_to_remove {
+            self.remove(txid);
+        }
+
+        // invalidate (roll back) any transparent spends in each invalidated tx
```

Comment:
Why not make invalidation of transparent utxos a method, that's analogous to the other pool-specific invalidation methods?

Reply:
okay, we will come back over this with another layer of polish